pkg/serviceaccounts/controllers/create_dockercfg_secrets.go

```diff
@@ -394,199 +442,192 @@ func (e *DockercfgController) syncServiceAccount(key string) error {
+func (e *DockercfgController) createDockerPullSecret(ctx context.Context, serviceAccount *v1.ServiceAccount) (*v1.Secret, bool, error) {
+	pendingTokenName := serviceAccount.Annotations[PendingTokenAnnotation]
+
+	// If this service account has no record of a pending token name, record one
+	if len(pendingTokenName) == 0 {
+		pendingTokenName = secret.Strategy.GenerateName(getDockercfgSecretNamePrefix(serviceAccount.Name))
+		if serviceAccount.Annotations == nil {
+			serviceAccount.Annotations = map[string]string{}
 		}
-		if !exists {
-			continue
+		serviceAccount.Annotations[PendingTokenAnnotation] = pendingTokenName
+		updatedServiceAccount, err := e.client.CoreV1().ServiceAccounts(serviceAccount.Namespace).Update(ctx, serviceAccount, metav1.UpdateOptions{})
+		// Conflicts mean we'll get called to sync this service account again
+		if kapierrors.IsConflict(err) {
+			return nil, false, nil
 		}
-		err = e.syncDockercfgOwner(secret.(*v1.Secret))
 		if err != nil {
-			return err
+			return nil, false, fmt.Errorf("failed to update the SA: %w", err)
 		}
+		serviceAccount = updatedServiceAccount
 	}
-	for _, secretRef := range serviceAccount.ImagePullSecrets {
-		secret, exists, err := e.secretCache.GetByKey(fmt.Sprintf("%s/%s", serviceAccount.Namespace, secretRef.Name))
-		if err != nil {
-			return err
+	currentSecret, err := e.client.CoreV1().Secrets(serviceAccount.Namespace).Get(ctx, pendingTokenName, metav1.GetOptions{})
+	if err != nil {
+		if kapierrors.IsNotFound(err) {
+			// this client supplies an empty struct in this case for some reason
+			currentSecret = nil
+		} else {
+			return nil, false, fmt.Errorf("failed to retrieve the current dockercfg secret: %w", err)
 		}
-		if !exists {
-			continue
+	}
+
+	if currentSecret != nil && len(currentSecret.Annotations[ServiceAccountTokenValueAnnotation]) != 0 {
+		return currentSecret, true, nil
+	}
+
+	if currentSecret == nil {
+		dockercfgSecret := &v1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      pendingTokenName,
+				Namespace: serviceAccount.Namespace,
+				Annotations: map[string]string{
+					v1.ServiceAccountNameKey: serviceAccount.Name,
+					v1.ServiceAccountUIDKey:  string(serviceAccount.UID),
+				},
+			},
+			Type: v1.SecretTypeDockercfg,
+			Data: map[string][]byte{
+				v1.DockerConfigKey: []byte("{}"), // required key but we have no data yet
+			},
+		}
+
+		blockDeletion := false
+		ownerRef := metav1.NewControllerRef(serviceAccount, v1.SchemeGroupVersion.WithKind("ServiceAccount"))
+		ownerRef.BlockOwnerDeletion = &blockDeletion
+		dockercfgSecret.SetOwnerReferences([]metav1.OwnerReference{*ownerRef})
+
+		klog.V(4).Infof("Creating dockercfg secret %q for service account %s/%s", dockercfgSecret.Name, serviceAccount.Namespace, serviceAccount.Name)
+
+		// Save the secret
+		_, err = e.client.CoreV1().Secrets(serviceAccount.Namespace).Create(ctx, dockercfgSecret, metav1.CreateOptions{})
```

Comment:
why create before getting the content?  With the tokenrequest API, the call to get the content is synchronous and it would simplify some of the waiting logic, right?